engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java

```diff
@@ -31,17 +32,30 @@ public void applyState(final long elementInstanceKey, final ProcessInstanceRecor
     elementInstanceState.newInstance(
         flowScopeInstance, elementInstanceKey, value, ProcessInstanceIntent.ELEMENT_ACTIVATING);
 
+    if (flowScopeInstance == null) {
+      // process instance level
+      return;
+    }
+
     final var flowScopeElementType = flowScopeInstance.getValue().getBpmnElementType();
     final var currentElementType = value.getBpmnElementType();
-    if (isContainerElement(flowScopeElementType, currentElementType)) {
-      // we currently spawn new tokens only for container elements
-      // we might spawn tokens for other bpmn element types as well later, then we can improve here
+
+    if (isContainerElement(flowScopeElementType, currentElementType)
+        || isNonInterruptingEventSubprocess(flowScopeInstance, currentElementType)
+        || currentElementType == BpmnElementType.BOUNDARY_EVENT
+        || currentElementType == BpmnElementType.INTERMEDIATE_CATCH_EVENT) {
```

Comment:
As already discussed with @korthout I'm not a fan of it but we currently not see a different approach. Maybe @saig0 has an idea?

Reply:
maybe we just remove the condition completely :thinking:

Reply:
Removing will not work, since we dont want to spawn tokens for interrupting event sub process but for boundary events. yolo

Reply:
Honestly to me it sounds like we will need some kind of dispatch mechanism like we have with the lifecycle for the commands. Probably not as a complex, but isn't it a bit what we want? 

```java
public final class ServiceTaskApplier implements ProcessInstanceApplier {
  @Override
  public void onActivating(final long key, final ProcessInstanceRecord record) {
    // do stuff
  }

  @Override
  public void onCompleting(final long key, final ProcessInstanceRecord record) {
    // do stuff
  }

  ...
}
```

Then this event applier becomes similar to `BpmnStreamProcessor`, I guess. Though tbh the wisest course of action here is probably to wait until we've migrated most or all, then see what the right abstraction is.

Reply:
Exactly. This is what I also commented there https://github.com/camunda-cloud/zeebe/issues/6497